pkg/tipselect/urts.go

```diff
@@ -0,0 +1,485 @@
+		},
+		// called on solid entry points
+		func(txHash hornet.Hash) {
+			updateIndexes(snapshotInfo.EntryPointIndex, snapshotInfo.EntryPointIndex)
+		}, true, false, nil); err != nil {
+		panic(err)
+	}
+
+	// update the outdated root snapshot indexes of all transactions in the cone in order from oldest txs to latest
+	ts.updateOutdatedRootSnapshotIndexes(outdatedTransactions, lsmi)
+
+	// set the new transaction root snapshot indexes in the metadata of the transaction
+	cachedTx.GetMetadata().SetRootSnapshotIndexes(youngestTxRootSnapshotIndex, oldestTxRootSnapshotIndex, lsmi)
+
+	return youngestTxRootSnapshotIndex, oldestTxRootSnapshotIndex
+}
+
+// calculateScore calculates the tip selection score of this transaction
+func (ts *TipSelector) calculateScore(txHash hornet.Hash, checkApprovees bool, lsmi milestone.Index) Score {
+	cachedTx := tangle.GetCachedTransactionOrNil(txHash) // tx +1
+	if cachedTx == nil {
+		panic(fmt.Sprintf("transaction not found: %v", txHash.Trytes()))
+	}
+	defer cachedTx.Release(true)
+
+	ytrsi, ortsi := ts.getTransactionRootSnapshotIndexes(cachedTx.Retain(), lsmi) // tx +1
+
+	// if the LSMI to YTRSI delta is over MaxDeltaTxYoungestRootSnapshotIndexToLSMI, then the tip is lazy
+	if (lsmi - ytrsi) > ts.maxDeltaTxYoungestRootSnapshotIndexToLSMI {
+		return ScoreLazy
+	}
+
+	// if the OTRSI to LSMI delta is over BelowMaxDepth/below-max-depth, then the tip is lazy
+	if (lsmi - ortsi) > ts.belowMaxDepth {
+		return ScoreLazy
+	}
+
+	if !checkApprovees {
+		// if we don't check for the approvee scores, this tip is not lazy.
+		// this is only used to verify that the tip is not lazy already when it gets picked.
+		return ScoreNonLazy
+	}
+
+	// the approvees (trunk and branch) are the transactions this tip approves
+	approveeHashes := map[string]struct{}{
+		string(cachedTx.GetTransaction().GetTrunkHash()):  {},
+		string(cachedTx.GetTransaction().GetBranchHash()): {},
+	}
+
+	approveesLazy := 0
+	for approveeHash := range approveeHashes {
+		approveeScore := ts.calculateScore(hornet.Hash(approveeHash), true, lsmi)
+
+		// direct approvee is already lazy, therefore so is this tip
+		if approveeScore == ScoreLazy {
+			return ScoreLazy
+		}
+
+		cachedApproveeTx := tangle.GetCachedTransactionOrNil(hornet.Hash(approveeHash)) // tx +1
+		if cachedApproveeTx == nil {
+			panic(fmt.Sprintf("transaction not found: %v", hornet.Hash(approveeHash).Trytes()))
+		}
+
+		_, approveeORTSI := ts.getTransactionRootSnapshotIndexes(cachedApproveeTx.Retain(), lsmi) // tx +1
+		cachedApproveeTx.Release(true)
+
+		// if the OTRSI to LSMI delta of the approvee is MaxDeltaTxApproveesOldestRootSnapshotIndexToLSMI, we mark it as such
+		if lsmi-approveeORTSI > ts.maxDeltaTxApproveesOldestRootSnapshotIndexToLSMI {
+			approveesLazy++
+		}
+	}
+
+	// if all available approvees' OTRSI violates the LSMI delta in relation to C2 the tip is lazy too
+	if len(approveeHashes) == approveesLazy {
```

Comment:
I like you using len here instead of hardcoded 2 as in the RFC, because this way you handle trunk == branch 👍

Reply:
Exactly, RFC is not optimal in that case :)